src/main/commandServer/settingsValidator.ts

```diff
@@ -0,0 +1,234 @@
+    switch (desiredEngine) {
+    case 'containerd':
+    case 'moby':
+      break;
+    case 'docker':
+      desiredEngine = 'moby';
+      break;
+    default:
+      errors.push(`Invalid value for ${ fqname }: <${ desiredEngine }>; must be 'containerd', 'docker', or 'moby'`);
+
+      return false;
+    }
+
+    return currentValue !== desiredEngine;
+  }
+
+  protected checkEnabled(currentState: boolean, desiredState: string|boolean, errors: string[], fqname: string): boolean {
+    if (typeof (desiredState) !== 'boolean') {
+      switch (desiredState) {
+      case 'true':
+        desiredState = true;
+        break;
+      case 'false':
+        desiredState = false;
+        break;
+      default:
+        errors.push(`Invalid value for ${ fqname }: <${ desiredState }>`);
+
+        return false;
+      }
+    }
+
+    return currentState !== desiredState;
+  }
+
+  protected checkKubernetesVersion(currentValue: string, desiredVersion: string, errors: string[], fqname: string): boolean {
+    const ptn = /^v?(\d+\.\d+\.\d+)(?:\+k3s\d+)?$/;
+    const m = ptn.exec(desiredVersion);
+
+    if (!m) {
+      errors.push(`Desired kubernetes version not valid: <${ desiredVersion }>`);
+
+      return false;
+    }
+    desiredVersion = m[1];
+    if (this.k8sVersions.length === 0) {
+      errors.push(`Can't check field ${ fqname }: no versions of Kubernetes were found.`);
+
+      return false;
+    } else if (!this.k8sVersions.includes(desiredVersion)) {
+      errors.push(`Kubernetes version ${ desiredVersion } not found.`);
+
+      return false;
+    }
+
+    return currentValue !== desiredVersion;
+  }
+
+  protected notSupported(fqname: string) {
+    return `Changing field ${ fqname } via the API isn't supported.`;
+  }
+
+  protected checkUnchanged(currentValue: any, desiredValue: any, errors: string[], fqname: string): boolean {
+    if (currentValue !== desiredValue) {
+      errors.push(this.notSupported(fqname));
+    }
+
+    return false;
+  }
+
+  // only arrays support stringification, so convert objects to arrays of tuples and sort on the keys
+  protected stableSerializeWSLIntegrations(value: Record<string, boolean>) {
+    return JSON.stringify(Object.entries(value).sort());
+  }
+
+  protected checkWSLIntegrations(currentValue: Record<string, boolean>, desiredValue: Record<string, boolean>, errors: string[], fqname: string): boolean {
+    if (typeof (desiredValue) !== 'object') {
+      errors.push(`Proposed field ${ fqname } should be an object, got <${ desiredValue }>.`);
+
+      return false;
+    }
+    try {
+      if (this.stableSerializeWSLIntegrations(currentValue) !== this.stableSerializeWSLIntegrations(desiredValue)) {
+        errors.push(this.notSupported(fqname));
+      }
+    } catch (err) {
+      errors.push(`JSON-parsing error checking field ${ fqname }: ${ err }`);
+    }
+
+    return false;
+  }
+
+  correctSynonymValues(newSettings: settingsLike): void {
```

Comment:
Seems like it would be easier to call this at the start of `checkProposedSettings` (so the validation doesn't have to deal with this), but as long as it works…

Reply:
That would simplify the code a bit. I'll probably come up with a PR to do this after.